modules/references/tests/server/reference-create.server.routes.tests.js

```diff
@@ -0,0 +1,871 @@
+'use strict';
+
+var should = require('should'),
+    request = require('supertest'),
+    async = require('async'),
+    path = require('path'),
+    sinon = require('sinon'),
+    mongoose = require('mongoose'),
+    User = mongoose.model('User'),
+    Reference = mongoose.model('Reference'),
+    testutils = require(path.resolve('./testutils/server.testutil')),
+    express = require(path.resolve('./config/lib/express'));
+
+describe('Create a reference', function () {
+
+  // user can leave a reference to anyone
+  //  - types of interaction
+  //  - recommend
+  //  - from whom
+  //  - to whom
+  // POST /references
+  // reference can't be modified or removed
+  // email notification will be sent to the receiver of the reference
+  // the receiver has some time to give a reference, too.
+  // after this time the only accepted answers are yes/ignore.
+  // after the given time or after both left reference, both references become public
+
+  // we'll catch email and push notifications
+  var jobs = testutils.catchJobs();
+
+  var user1,
+      user2,
+      user3Nonpublic;
+
+  var app = express.init(mongoose.connection);
+  var agent = request.agent(app);
+
+  var _user1 = {
+    public: true,
+    firstName: 'Full',
+    lastName: 'Name',
+    displayName: 'Full Name',
+    email: 'user1@example.com',
+    username: 'user1',
+    displayUsername: 'user1',
+    password: 'correcthorsebatterystaples',
+    provider: 'local'
+  };
+
+  var _user2 = {
+    public: true,
+    firstName: 'Full2',
+    lastName: 'Name2',
+    displayName: 'Full2 Name2',
+    email: 'user2@example.com',
+    username: 'user2',
+    displayUsername: 'user2',
+    password: 'correcthorsebatterystaples',
+    provider: 'local'
+  };
+
+  var _user3Nonpublic = {
+    public: false,
+    firstName: 'Full3',
+    lastName: 'Name3',
+    displayName: 'Full3 Name3',
+    email: 'user3@example.com',
+    username: 'user3',
+    displayUsername: 'user3',
+    password: 'correcthorsebatterystaples',
+    provider: 'local'
+  };
+
+  beforeEach(function () {
+    sinon.useFakeTimers({ now: 1500000000000, toFake: ['Date'] });
+  });
+
+  afterEach(function () {
+    sinon.restore();
+  });
+
+  beforeEach(function (done) {
+
+    user1 = new User(_user1);
+    user2 = new User(_user2);
+    user3Nonpublic = new User(_user3Nonpublic);
+
+    async.eachSeries([user1, user2, user3Nonpublic], function (user, cb) {
+      user.save(cb);
+    }, done);
+  });
+
+  afterEach(function (done) {
+    Reference.deleteMany().exec(function () {
+      User.deleteMany().exec(done);
+    });
+  });
+
+  /**
+   * @TODO this can be refactored. Sign in may be moved to some test utils
```

Comment:
Maybe can do it? ;)

Reply:
It is done in a later PR. I refactored this when needed to reuse.

https://github.com/Trustroots/trustroots/pull/759/files#diff-4166014cc218ad965962004b2f8b9f31R136